src/eos-updater-poll-common.c

```diff
@@ -386,41 +391,37 @@ get_refspec_to_upgrade_on (gchar               **refspec_to_upgrade_on,
   if (booted_deployment == NULL)
     return FALSE;
 
-  if (!get_refspec_to_upgrade_on_from_deployment (sysroot,
-                                                  booted_deployment,
-                                                  &refspec,
-                                                  error))
+  if (!get_booted_refspec (booted_deployment,
+                           &refspec,
+                           &remote,
+                           &ref,
+                           &collection_ref,
+                           error))
     return FALSE;
 
-  /* Nothing here */
-  if (refspec == NULL)
-    return get_booted_refspec (booted_deployment,
-                               refspec_to_upgrade_on,
-                               remote_to_upgrade_on,
-                               ref_to_upgrade_on,
-                               collection_ref_to_upgrade_on,
-                               error);
-
-  /* We'll need the repo to get the collection-id */
-  if (!ostree_sysroot_get_repo (sysroot, &repo, NULL, error))
+  if (!get_ref_to_upgrade_on_from_deployment (sysroot,
+                                              booted_deployment,
+                                              &ref_for_deployment,
+                                              error))
     return FALSE;
 
-  /* Found a refspec on this commit's metadata */
-  if (!ostree_parse_refspec (refspec, &remote, &ref, error))
-    return FALSE;
-  if (remote == NULL)
+  /* Handle the ref from the commit's metadata */
+  if (ref_for_deployment != NULL)
     {
-      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
-                   "Invalid refspec ‘%s’ in commit: did not contain a remote name",
-                   refspec);
-      return FALSE;
+      OstreeCollectionRef *orig_collection_ref = collection_ref;
```

Comment:
Instead of freeing this later you can take ownership with `g_autoptr`

Reply:
Ah, good call. I always think of `g_autoptr` as being related to allocation rather than an automatic finalizer.